src/property.ts

```diff
@@ -4,252 +4,188 @@
+    }
+    return inputValue instanceof Array
+        ? result
+        : result[0];
+};
+
+export const createPropertyDefinition = (opt: PropertyDefinitionInput): PropertyDefinition => {
+    const {
+        type: inputType,
+        default: inputDefault,
+        ...rest
+    } = opt;
+    const defaultType = ((opt.min !== undefined || opt.max !== undefined) && !opt.type)
+        ? 'integer'
+        : 'string';
+    const type: DbType = inputType || defaultType;
+
+    let defaultCast;
+
+    if (!opt.cast) { // set the default util.cast functions
+        if (type === 'integer') {
+            defaultCast = util.castInteger;
+        } else if (type === 'string') {
+            if (!opt.nullable) {
+                defaultCast = (opt.nonEmpty
+                    ? util.castLowercaseNonEmptyString
+                    : util.castLowercaseString);
+            } else {
+                defaultCast = (opt.nonEmpty
+                    ? util.castLowercaseNonEmptyNullableString
+                    : util.castLowercaseNullableString);
+            }
+        } else if (type.includes('link')) {
+            if (!opt.nullable) {
+                defaultCast = util.castToRID;
+            } else {
+                defaultCast = util.castNullableLink;
+            }
         }
-        if ((prop.maxItems || prop.maxItems === 0) && result.length > prop.maxItems) {
-            throw new ValidationError(`Violated the maxItems constraint of ${prop.name}. More than the allowed number of elements (${result.length} > ${prop.maxItems})`);
+    }
+
+    let generateDefault,
+        defaultValue;
+
+    if (opt.default !== undefined) {
+        if (opt.default instanceof Function) {
+            generateDefault = opt.default;
+        } else {
+            defaultValue = opt.default;
         }
-        return inputValue instanceof Array
-            ? result
-            : result[0];
     }
 
-    /**
-     * Given some value for a property, ensure that it does not violate the current model contraints
-     *
-     * @throws {ValidationError} if the input value violates any of the property model constraints
-     *
-     * @returns the cast input value
-     */
-    validate(inputValue) {
-        return Property.validateWith(this, inputValue);
+    const castFunction = opt.cast || defaultCast;
+    let choices;
+
+    if (castFunction && opt.choices) {
+        choices = opt.choices.map((choice) => castFunction(choice));
     }
-}
+
+    const result: PropertyDefinition = {
+        ...rest,
+        type,
+        cast: castFunction,
+        default: defaultValue,
+        generateDefault,
+        description: opt.description || '',
+        generated: Boolean(opt.generated),
+        generationDependencies: Boolean(opt.generationDependencies),
+        iterable: Boolean(/(set|list|bag|map)/ig.exec(type)),
+        nullable: (
```

Comment:
non-blocking:
as long as it's not a case where you don't actually have to make a distinction between null/undefined, I find using `??` easier to read

for this case it would change
```ts
opt.nullable === undefined
    ? true
    : opt.nullable
```
to
```ts
opt.nullable ?? true
```

Reply:
In this case I specifically want to know that the user didn't specify anything, whereas were they to pass null i'd expect this to default to false